src/file/file.repository.ts

```diff
@@ -20,61 +22,141 @@ const sort = {
+   * @param file - the file that will be converted.
+   */
+  static populatedShortcutToFile(file: IPopulatedShortcut): IFile {
+    const shortcutAsFile: IFile = { ...file.fileID, ...file };
+    delete shortcutAsFile.fileID;
+    return shortcutAsFile;
+  }
+
+  /**
+   * Get a baseFile type file and converts its type to IFile.
+   * @param file - the file that will be converted.
+   */
+  static async baseFileToIFile(file: any): Promise<IFile> {
+    const factoryFile: PrimitiveFile = this.fileFactory(file, file.fileModel);
+    let responseFile: IFile = factoryFile as IFile;
+    if (factoryFile instanceof IShortcut) {
+      const populatedShortcut: IPopulatedShortcut = await (await file.populate('fileID').execPopulate()).toObject();
+      responseFile = this.populatedShortcutToFile(populatedShortcut);
+    }
+    return responseFile
+  }
+
+  /**
+   * Get a file by its id.
    * @param id - the id of the file.
    */
-  static getById(id: string): Promise<IFile | null> {
-    return fileModel.findById({ _id: new ObjectID(id) }).exec();
+  static async getById(id: string): Promise<IFile> {
+    const file = await baseFileModel.findById({ _id: new ObjectID(id) }).exec();
+
+    return this.baseFileToIFile(file);
+  }
+
+  /**
+   * Get a file and its fileModel parameter and casts it to the correct class.
+   * @param file - the file that its type being casted
+   * @param type - the type that's using to cast the file
+   * @returns file that's being casted corrcetly
+   */
+  static fileFactory(file: any, type: string): PrimitiveFile {
+    switch (type) {
+      case fileModelName:
+        return new IFile(file);
+      case shortcutModelName:
+        return new IShortcut(file);
+      default:
+        throw new Error('File type not supported');
+    }
   }
 
   /**
-   * Get the file by its key.
+   * Get a file by its key.
    * @param key - the key of the file.
    */
-  static getByKey(key: string): Promise<IFile | null> {
-    return fileModel.findOne({ key }).exec();
+  static async getByKey(key: string): Promise<IFile> {
+    const file = await baseFileModel.findOne({ key }).exec();
+
+    return this.baseFileToIFile(file);
   }
 
   /**
-   * Get several filed by their ids.
-   * @param ids - the array of the ids.
+   * Get several files by their ids.
+   * @param ids - array of the files ids.
    */
-  static getByIds(ids: string[]): Promise<IFile[]> {
+  static async getByIds(ids: string[]): Promise<IFile[]> {
     const objIds: ObjectID[] = ids.map(id => new ObjectID(id));
-    return fileModel.find({
+    const files = await baseFileModel.find({
       _id: { $in: objIds },
     }).exec();
+    const settledFiles = await Promise.allSettled(files.map(this.baseFileToIFile));
+    const fulfilledFiles: IFile[] = (settledFiles.filter(file => file.status === 'fulfilled') as PromiseFulfilledResult<IFile>[]).map(file => file.value);
+    const rejectedFiles = settledFiles.filter(file => file.status === 'rejected');
```

Comment:
I think there is a better way to so using `reduce` maybe

Reply:
i can use reduce, but im pretty sure i talked to you about it when i did it and you agreed with me that filter is more efficient.
anyway, its done.

Reply:
i dont think its necessary, its not the perpose of reduce tbh.

Reply:
it makes things complicated tbh